### ISSUE TITLE
should JSON data classes implement equals() and hashCode() ?

### ISSUE THREAD

**Issue by HopingCoding** [stleary/JSON-java#63]
this is more a discussion, than an issue.

did you ever think about implementing the equals() and hashCode() functions in JSONObject, JSONArray and JSONString so that they recursively detect identical objects?

```
        JSONObject a = new JSONObject();
        a.put("key", "value");
        JSONObject b = new JSONObject();
        b.put("key", "value");

        a.equals(b); //would return true
```

for example with this, JSON data object could be used in java-collections to be checked in their contains() function.

**Comment by oconnelc** [comment]
Is there a reason this issue seems to be avoided? With Java Streams (which has been around since Java 1.8) equals and hashcode are useful for stream methods such as distinct.  If there isn't a reason why this shouldn't be done, would you take a pull request?

**Comment by stleary** [comment]
Hi @oconnelc, equals and hashcode would probably not be accepted at this time due to introducing a behavior change with no corresponding bug. Douglas Crockford later implemented [similar()](https://github.com/stleary/JSON-java/commit/7ff3fa4e40cfd5b44570297080810aed5cce6086#diff-6c8bf136596bcd17e616c3f3fb5654a4) for this lib. My guess is that similar() provides some of the benefits of equals() without incurring the potential overhead of hashcode() for deeply nested JSON docs.  Also, see #297

**Comment by oconnelc** [comment]
I figured it wouldn't be a small change.

One thing the similar does not provide is the ability to have JSONObjects
in any of Java's Colllections type as almost all of them use equals,
hashcode or both. As a result the only way to reliably use these Java types
is to create a wrapper object to contain the JSONObject or convert the
JSONObject to String and back.

For my use case I was able to use JSONArray's spliterator to make a stream
and then I wrote my own filter method to track uniqueness since the
distinct method fails due to a lack of an implemented equals.

Also I think it would help the community to have some sort of documentation
explaining why this isn't done. I found a couple of tickets asking for this
capability and they were all responded with "You have the code." To me that
sounds like I should go and submit a pull request. If this capabilty is
truely never going to be implemented, it should be documented as to why. I
also think projects like this shoukd encourage discussion and submitted
fixes, but I guess that's up to the owner. (Sorry if this turned into a
rant, I was hoping this could be resolved and I read the responses as
unhelpful.)
